core/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java

```diff
@@ -58,28 +57,26 @@ protected AbstractRestChannel(RestRequest request, boolean detailedErrorsEnabled
 
     @Override
     public XContentBuilder newBuilder() throws IOException {
-        return newBuilder(request.hasContent() ? request.content() : null, true);
+        return newBuilder(request.getXContentType(), true);
     }
 
     @Override
     public XContentBuilder newErrorBuilder() throws IOException {
         // Disable filtering when building error responses
-        return newBuilder(request.hasContent() ? request.content() : null, false);
+        return newBuilder(request.getXContentType(), false);
     }
 
     @Override
-    public XContentBuilder newBuilder(@Nullable BytesReference autoDetectSource, boolean useFiltering) throws IOException {
+    public XContentBuilder newBuilder(@Nullable XContentType xContentType, boolean useFiltering) throws IOException {
         XContentType contentType = XContentType.fromMediaTypeOrFormat(format);
         if (contentType == null) {
-            // try and guess it from the auto detect source
-            if (autoDetectSource != null) {
-                contentType = XContentFactory.xContentType(autoDetectSource);
+            if (xContentType != null) {
+                contentType = xContentType;
```

Comment:
ok, I think I missed that the content-type argument will be auto-detected by default, or optionally read from Content-Type instead. Then we are 100% compatible by default, you are right.